File: spec/test-lsp.hs
{-# LANGUAGE OverloadedStrings #-}

import           Test.Hspec
import           Control.Monad.IO.Class
import           Control.Lens
import           Data.Text
import           Language.LSP.Test
import           Language.LSP.Types
import           Language.LSP.Types.Lens

digestif_cmd = "digestif"
fixtures = "spec/fixtures"

mySession = runSession digestif_cmd fullCaps fixtures

main = hspec $ do
  describe "textDocument/completion" $ do
    it "completes commands" $ mySession $ do
      doc <- openDoc "file1.tex" "latex"
      item:_ <- getCompletions doc (Position 3 7)
      liftIO $ do
        item ^. label `shouldBe` "raisebox"
        item ^. detail `shouldBe` Just "{distance}[height][depth]{text}"

    it "completes commands from a package" $ mySession $ do
      doc <- openDoc "using-tikz.tex" "latex"
      item:_ <- getCompletions doc (Position 1 8)
      --items <- getCompletions doc (Position 1 8)
      liftIO $ item ^. label `shouldBe` "pgfmatrix"

    it "completes tikz options" $ mySession $ do
      doc <- openDoc "using-tikz.tex" "latex"
      env_opt:_ <- getCompletions doc (Position 3 40)
      path_opt:_ <- getCompletions doc (Position 4 18)
      liftIO $ do
        env_opt ^. label `shouldBe` "execute at begin picture"
        path_opt ^. label `shouldBe` "line cap"
        path_opt ^. detail `shouldBe` Just "=⟨type⟩"

    it "completes local labels" $ mySession $ do
      doc <- openDoc "file1.tex" "latex"
      item:_ <- getCompletions doc (Position 5 7)
      liftIO $ item ^. label `shouldSatisfy` isPrefixOf "somelabel"

    it "completes label from included file" $ mySession $ do
      doc <- openDoc "root.tex" "latex"
      item:_ <- getCompletions doc (Position 2 8)
      liftIO $ item ^. label `shouldSatisfy` isPrefixOf "childlabel"

    it "completes label from root file" $ mySession $ do
      doc <- openDoc "child.tex" "latex"
      item:_ <- getCompletions doc (Position 2 8)
      liftIO $ item ^. label `shouldSatisfy` isPrefixOf "rootlabel"

    it "completes citation from thebibliography environment" $ mySession $ do
      doc <- openDoc "file1.tex" "latex"
      item:_ <- getCompletions doc (Position 9 10)
      liftIO $ item ^. label `shouldSatisfy` isPrefixOf "somebibitem"

    it "completes citation from bibtex file" $ mySession $ do
      doc <- openDoc "child.tex" "latex"
      item:_ <- getCompletions doc (Position 3 8)
      liftIO $ do
        item ^. label `shouldSatisfy` isPrefixOf "somepaper"
        item ^. detail `shouldBe` Nothing

    it "completes label wth fuzzy match" $ mySession $ do
      doc <- openDoc "file2.tex" "latex"
      item:_ <- getCompletions doc (Position 2 10)
      liftIO $ item ^. label `shouldSatisfy` isPrefixOf "somelabel"

    it "completes citation wth fuzzy match" $ mySession $ do
      doc <- openDoc "file2.tex" "latex"
      item:_ <- getCompletions doc (Position 10 10)
      liftIO $ item ^. label `shouldSatisfy` isPrefixOf "somepaper"

  describe "textDocument/completion with inconsistent root information" $ do
    it "works when root doesn't exist" $ mySession $ do
      doc <- openDoc "child.tex" "latex"
      let changeEv = TextDocumentContentChangeEvent (Just (Range (Position 0 14) (Position 0 18))) (Just 4) "nonexistent"
      changeDoc doc [changeEv]
      items <- getCompletions doc (Position 2 8)
      liftIO $ items `shouldBe` []

    it "works when root doesn't exist" $ mySession $ do
      doc <- openDoc "child.tex" "latex"
      let changeEv = TextDocumentContentChangeEvent (Just (Range (Position 0 14) (Position 0 18))) (Just 4) "nonexistent"
      changeDoc doc [changeEv]
      item:_ <- getCompletions doc (Position 4 8)
      liftIO $ item ^. label `shouldSatisfy` isPrefixOf "childlabel"

    it "works when root doesn't refer back to child" $ mySession $ do
      doc <- openDoc "child.tex" "latex"
      let edit = TextEdit (Range (Position 0 14) (Position 0 18)) "file1"
      changeDoc doc [TextDocumentContentChangeEvent (Just (Range (Position 0 14) (Position 0 18))) (Just 4) "file1"]
      item:_ <- getCompletions doc (Position 4 8)
      liftIO $ item ^. label `shouldSatisfy` isPrefixOf "childlabel"

  describe "textDocument/definition" $
    it "works" $ mySession $ do
      doc <- openDoc "file1.tex" "latex"
      def <- getDefinitions doc (Position 6 6)
      liftIO $ def `shouldBe` (InL [Location (doc ^. uri) (Range (Position 4 7) (Position 4 16))])

  describe "textDocument/references" $ do
    it "finds references in a single file" $ mySession $ do
      let pos = (Position 6 6)
      doc <- openDoc "file1.tex" "latex"
      defs <- getReferences doc pos True
      liftIO $ defs `shouldBe` (Language.LSP.Types.List [
        Location (doc ^. uri) (Range (Position 4 7) (Position 4 16)),
        Location (doc ^. uri) (Range (Position 7 5) (Position 7 14))])

    it "finds references across files" $ mySession $ do
      doc <- openDoc "root.tex" "latex"
      doc2 <- openDoc "child.tex" "latex"
      defs <- getReferences doc (Position 3 13) True
      liftIO $ defs `shouldBe` (Language.LSP.Types.List [
        Location (doc2 ^. uri) (Range (Position 1 7) (Position 1 17)),
        Location (doc ^. uri) (Range (Position 3 5) (Position 3 15))])

  describe "textDocument/hover" $ do
    it "finds command documentation" $ mySession $ do
      let getText (HoverContents (MarkupContent MkMarkdown x)) = x
      doc <- openDoc "file2.tex" "latex"
      Just h <- getHover doc (Position 1 5)
      liftIO $ getText(h ^. contents) `shouldSatisfy` isInfixOf "do nothing"

    it "finds context around a label" $ mySession $ do
      let getText (HoverContents (MarkupContent MkMarkdown x)) = x
      doc <- openDoc "file2.tex" "latex"
      Just h <- getHover doc (Position 0 5)
      -- liftIO $ print $ getText(h ^. contents)
      liftIO $ do
        getText(h ^. contents) `shouldSatisfy` isInfixOf "Lorem ipsum"
        getText(h ^. contents) `shouldSatisfy` isInfixOf "minim veniam"

File: spec/test_lsp.py
from pathlib import Path
from urllib.parse import urlparse
from urllib.request import url2pathname

import pytest
import pytest_lsp
from lsprotocol.types import (
    Location,
    MarkupKind,
    Position,
    Range,
    ReferenceContext,
    ReferenceParams,
    TextDocumentIdentifier,
)
from pytest_lsp import ClientServerConfig
from pytest_lsp.client import LanguageClient

BIN_DIR = Path(__file__).parents[1].joinpath("bin")
FIXTURE_DIR = Path(__file__).parent.joinpath("fixtures")


@pytest_lsp.fixture(
    config=ClientServerConfig(
        server_command=["lua", BIN_DIR / "digestif", "-v"],
        root_uri=FIXTURE_DIR.as_uri(),
    ),
)
async def client():
    pass


def open_doc(client, file, language):
    root = Path(url2pathname(urlparse(client.root_uri).path))
    doc = root.joinpath(file)
    with open(doc) as f:
        contents = f.read()
    client.notify_did_open(doc.as_uri(), language, contents)
    return doc.as_uri()


# Basic completion tests


@pytest.mark.asyncio
async def test_completion_latex(client: LanguageClient):
    "Complete a built-in command."
    doc_uri = open_doc(client, "file1.tex", "latex")
    result = await client.completion_request(doc_uri, 3, 7)
    assert result[0].label == "raisebox"
    assert result[0].detail == "{distance}[height][depth]{text}"
    assert result[0].documentation


@pytest.mark.asyncio
async def test_completion_from_package(client: LanguageClient):
    "Complete a command loaded from a package."
    doc_uri = open_doc(client, "using-tikz.tex", "latex")
    result = await client.completion_request(doc_uri, 1, 8)
    assert result[0].label == "pgfmatrix"


@pytest.mark.asyncio
async def test_completion_tikz_option(client: LanguageClient):
    "Complete a TikZ option."
    doc_uri = open_doc(client, "using-tikz.tex", "latex")
    result = await client.completion_request(doc_uri, 3, 40)
    assert result[0].label == "execute at begin picture"
    result = await client.completion_request(doc_uri, 4, 18)
    assert result[0].label == "line cap"
    assert result[0].detail == "=⟨type⟩"


@pytest.mark.asyncio
async def test_completion_label(client: LanguageClient):
    "Complete a label defined in the current file."
    doc_uri = open_doc(client, "file1.tex", "latex")
    result = await client.completion_request(doc_uri, 5, 7)
    assert result[0].label.startswith("somelabel")
    assert result[0].text_edit.new_text == "somelabel"


@pytest.mark.asyncio
async def test_completion_label_from_child(client: LanguageClient):
    "Complete a label defined in an included file, referenced in the root file."
    doc_uri = open_doc(client, "root.tex", "latex")
    result = await client.completion_request(doc_uri, 2, 8)
    assert result[0].label.startswith("childlabel")


@pytest.mark.asyncio
async def test_completion_label_from_root(client: LanguageClient):
    "Complete a label defined in the root file, referenced in an included file."
    doc_uri = open_doc(client, "child.tex", "latex")
    result = await client.completion_request(doc_uri, 2, 8)
    assert result[0].label.startswith("rootlabel")


@pytest.mark.asyncio
async def test_completion_thebibliography(client: LanguageClient):
    "Complete a citation defined in 'thebibliogprahy' environment."
    doc_uri = open_doc(client, "file1.tex", "latex")
    result = await client.completion_request(doc_uri, 9, 10)
    assert result[0].label.startswith("somebibitem")
    assert result[0].text_edit.new_text == "somebibitem"


@pytest.mark.asyncio
async def test_completion_bibtex(client: LanguageClient):
    "Complete a citation defined in a bibtex file."
    doc_uri = open_doc(client, "child.tex", "latex")
    result = await client.completion_request(doc_uri, 3, 8)
    assert result[0].label.startswith("somepaper")
    assert result[0].text_edit.new_text == "somepaper"
    assert result[0].detail is None


@pytest.mark.asyncio
async def test_completion_citation_fuzzy(client: LanguageClient):
    "Complete a citation defined in 'thebibliogprahy' environment."
    doc_uri = open_doc(client, "file2.tex", "latex")
    result = await client.completion_request(doc_uri, 10, 10)
    assert result[0].label.startswith("somepaper")
    assert result[0].text_edit.new_text == "somepaper"


# Completion with inconsistent root information


@pytest.mark.asyncio
async def test_completion_no_root(client: LanguageClient):
    "Try to complete locally when root document doesn't exist."
    doc_uri = open_doc(client, "child.tex", "latex")
    client.notify_did_change(doc_uri, text="xxxx", line=0, character=14)
    result = await client.completion_request(doc_uri, 4, 8)
    assert result[0].label.startswith("childlabel")


@pytest.mark.asyncio
async def test_completion_broken_root(client: LanguageClient):
    "Try to complete locally when root doesn't refer back to child."
    doc_uri = open_doc(client, "child.tex", "latex")
    client.notify_did_change(doc_uri, text="file1.tex\n", line=0, character=14)
    result = await client.completion_request(doc_uri, 4, 8)
    assert result[0].label.startswith("childlabel")


@pytest.mark.asyncio
async def test_completion_from_root_no_root(client: LanguageClient):
    "Try to complete label from root document when it doesn't exist."
    doc_uri = open_doc(client, "child.tex", "latex")
    result = await client.completion_request(doc_uri, 2, 8)
    assert result[0].label.startswith("rootlabel")
    client.notify_did_change(doc_uri, text="xxxx", line=0, character=14)
    result = await client.completion_request(doc_uri, 2, 8)
    assert result == []


# Tests for textDocument/definition method


@pytest.mark.asyncio
async def test_definition(client: LanguageClient):
    "Find the definition Try to complete locally when root doesn't refer back to child."
    doc_uri = open_doc(client, "file1.tex", "latex")
    result = await client.definition_request(doc_uri, 7, 6)
    assert result == Location(
        uri=doc_uri,
        range=Range(
            start=Position(line=4, character=7), end=Position(line=4, character=16)
        ),
    )


# Tests for textDocument/references method


@pytest.mark.asyncio
async def test_references(client: LanguageClient):
    "Find references in a single file."
    doc_uri = open_doc(client, "file1.tex", "latex")
    result = await client.text_document_references_request(
        ReferenceParams(
            context=ReferenceContext(include_declaration=True),
            text_document=TextDocumentIdentifier(uri=doc_uri),
            position=Position(line=7, character=6),
        )
    )
    assert len(result) == 2
    assert result[0] == Location(
        uri=doc_uri,
        range=Range(
            start=Position(line=4, character=7), end=Position(line=4, character=16)
        ),
    )
    assert result[1] == Location(
        uri=doc_uri,
        range=Range(
            start=Position(line=7, character=5), end=Position(line=7, character=14)
        ),
    )


@pytest.mark.asyncio
async def test_references_across_files(client: LanguageClient):
    "Find references across file."
    root_uri = open_doc(client, "root.tex", "latex")
    child_uri = open_doc(client, "child.tex", "latex")
    result = await client.text_document_references_request(
        ReferenceParams(
            context=ReferenceContext(include_declaration=True),
            text_document=TextDocumentIdentifier(uri=root_uri),
            position=Position(line=3, character=13),
        )
    )
    assert len(result) == 2
    assert result[0] == Location(
        uri=child_uri,
        range=Range(
            start=Position(line=1, character=7), end=Position(line=1, character=17)
        ),
    )
    assert result[1] == Location(
        uri=root_uri,
        range=Range(
            start=Position(line=3, character=5), end=Position(line=3, character=15)
        ),
    )


# Tests for textDocument/hover method


@pytest.mark.asyncio
async def test_hover_command(client: LanguageClient):
    "Find command documentation."
    doc_uri = open_doc(client, "file2.tex", "latex")
    result = await client.hover_request(doc_uri, 1, 5)
    assert result.contents.kind == MarkupKind.Markdown
    assert result.contents.value.startswith("`\\relax`:")
    assert "do nothing" in result.contents.value


@pytest.mark.asyncio
async def test_hover_label(client: LanguageClient):
    "Find context around a label."
    doc_uri = open_doc(client, "file2.tex", "latex")
    result = await client.hover_request(doc_uri, 0, 5)
    assert result.contents.kind == MarkupKind.Markdown
    assert result.contents.value.startswith("`somelabel`:")
    assert "Lorem ipsum" in result.contents.value
    assert "minim veniam" in result.contents.value
